### PR TITLE
Bug 2075475: allow creating cluster subnet route for egress router pod

### PR DESCRIPTION
**Pull request by msherif1234**
egress router pod doesn't have route to allow this pod to connect
to other pods on different nodes.

to be able to reach svc backed by egress router pod from pods
on different node we need to install this route

unit-test

started cluster-bot with this PR and checked the routing table inside egress router pod after the changes
[root@ci-ln-4gkcnf2-72292-zxd7p-worker-a-7p29w /]# ip r
default via 192.168.123.1 dev net1 
10.128.0.0/14 via 10.131.0.1 dev eth0 <<<<
10.131.0.0/23 dev eth0 proto kernel scope link src 10.131.0.20 
172.30.0.0/16 via 10.131.0.1 dev eth0 <<<<
192.168.123.0/24 dev net1 proto kernel scope link src 192.168.123.144 
192.168.123.1 dev net1 

```
oc get pod egress-router-cni-deployment-6f8c4647d6-gbjv6  -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    k8s.ovn.org/pod-networks: '{"default":{"ip_addresses":["10.131.0.20/23"],"mac_address":"0a:58:0a:83:00:14","routes":[{"dest":"10.128.0.0/14","nextHop":"10.131.0.1"},{"dest":"172.30.0.0/16","nextHop":"10.131.0.1"}],"ip_address":"10.131.0.20/23"}}'
```

Signed-off-by: Mohamed Mahmoud <mmahmoud@redhat.com>

### CODE REVIEW

**Comment by msherif1234** [comment]
> I don't think that this is the correct component for this change. If we look at the egress router pod definition, we see that it has a default route towards the k8s network and no route at all towards the outside:
> 
> ```
> [root@openshift-jumpserver-0 ~]# oc get pod egress-router-cni-deployment-575dc66589-c58kh -o yaml
> apiVersion: v1
> kind: Pod
> metadata:
>   annotations:
>     k8s.ovn.org/pod-networks: '{"default":{"ip_addresses":["172.24.3.89/23"],"mac_address":"0a:58:ac:18:03:59","gateway_ips":["172.24.2.1"],"ip_address":"172.24.3.89/23","gateway_ip":"172.24.2.1"}}'
>     k8s.v1.cni.cncf.io/network-status: |-
>       [{
>           "name": "ovn-kubernetes",
>           "interface": "eth0",
>           "ips": [
>               "172.24.3.89"
>           ],
>           "mac": "0a:58:ac:18:03:59",
>           "default": true,
>           "dns": {}
>       },{
>           "name": "sriov-testing/egress-router-cni-nad",
>           "interface": "net1",
>           "ips": [
>               "192.168.123.180"
>           ],
>           "mac": "c2:37:f4:c0:b7:f0",
>           "dns": {}
>       }]
>     k8s.v1.cni.cncf.io/networks: egress-router-cni-nad
>     k8s.v1.cni.cncf.io/networks-status: |-
>       [{
>           "name": "ovn-kubernetes",
>           "interface": "eth0",
>           "ips": [
>               "172.24.3.89"
>           ],
>           "mac": "0a:58:ac:18:03:59",
>           "default": true,
>           "dns": {}
>       },{
>           "name": "sriov-testing/egress-router-cni-nad",
>           "interface": "net1",
>           "ips": [
>               "192.168.123.180"
>           ],
>           "mac": "c2:37:f4:c0:b7:f0",
>           "dns": {}
>       }]
>     openshift.io/scc: restricted
>   creationTimestamp: "2022-04-21T14:43:29Z"
>   generateName: egress-router-cni-deployment-575dc66589-
>   labels:
>     app: egress-router-cni
>     pod-template-hash: 575dc66589
>   name: egress-router-cni-deployment-575dc66589-c58kh
>   namespace: sriov-testing
>   ownerReferences:
>   - apiVersion: apps/v1
>     blockOwnerDeletion: true
>     controller: true
>     kind: ReplicaSet
>     name: egress-router-cni-deployment-575dc66589
>     uid: 8d4b3cee-3a1c-4cd0-ac7b-b63ce1c82fcd
>   resourceVersion: "1050704"
>   uid: 334c51c2-3a3f-4705-a8d8-5c8481541a16
> spec:
>   containers:
>   - command:
>     - /bin/sh
>     - -c
>     - sleep infinity
>     image: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:eab979221383ee77cb9ddd170076006c5e4b7eda9dd214a0c4ddd09c660411dd
>     imagePullPolicy: IfNotPresent
>     name: egress-router-cni-pod
>     resources: {}
>     securityContext:
>       capabilities:
>         drop:
>         - KILL
>         - MKNOD
>         - SETGID
>         - SETUID
>       runAsUser: 1000670000
>     terminationMessagePath: /dev/termination-log
>     terminationMessagePolicy: FallbackToLogsOnError
>     volumeMounts:
>     - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
>       name: kube-api-access-qvtr7
>       readOnly: true
>   dnsPolicy: ClusterFirst
>   enableServiceLinks: true
>   imagePullSecrets:
>   - name: default-dockercfg-n2p9q
>   nodeName: openshift-worker-1
>   preemptionPolicy: PreemptLowerPriority
>   priority: 0
>   restartPolicy: Always
>   schedulerName: default-scheduler
>   securityContext:
>     fsGroup: 1000670000
>     seLinuxOptions:
>       level: s0:c26,c10
>   serviceAccount: default
>   serviceAccountName: default
>   terminationGracePeriodSeconds: 30
>   tolerations:
>   - effect: NoExecute
>     key: node.kubernetes.io/not-ready
>     operator: Exists
>     tolerationSeconds: 300
>   - effect: NoExecute
>     key: node.kubernetes.io/unreachable
>     operator: Exists
>     tolerationSeconds: 300
>   volumes:
>   - name: kube-api-access-qvtr7
>     projected:
>       defaultMode: 420
>       sources:
>       - serviceAccountToken:
>           expirationSeconds: 3607
>           path: token
>       - configMap:
>           items:
>           - key: ca.crt
>             path: ca.crt
>           name: kube-root-ca.crt
>       - downwardAPI:
>           items:
>           - fieldRef:
>               apiVersion: v1
>               fieldPath: metadata.namespace
>             path: namespace
>       - configMap:
>           items:
>           - key: service-ca.crt
>             path: service-ca.crt
>           name: openshift-service-ca.crt
> status:
>   conditions:
>   - lastProbeTime: null
>     lastTransitionTime: "2022-04-21T14:43:25Z"
>     status: "True"
>     type: Initialized
>   - lastProbeTime: null
>     lastTransitionTime: "2022-04-21T14:43:29Z"
>     status: "True"
>     type: Ready
>   - lastProbeTime: null
>     lastTransitionTime: "2022-04-21T14:43:29Z"
>     status: "True"
>     type: ContainersReady
>   - lastProbeTime: null
>     lastTransitionTime: "2022-04-21T14:43:28Z"
>     status: "True"
>     type: PodScheduled
>   containerStatuses:
>   - containerID: cri-o://d76464b7dadb492e71880b3e8bd2e47d9e20fbba65de4169b6d71ea20222e6d8
>     image: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:eab979221383ee77cb9ddd170076006c5e4b7eda9dd214a0c4ddd09c660411dd
>     imageID: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:eab979221383ee77cb9ddd170076006c5e4b7eda9dd214a0c4ddd09c660411dd
>     lastState: {}
>     name: egress-router-cni-pod
>     ready: true
>     restartCount: 0
>     started: true
>     state:
>       running:
>         startedAt: "2022-04-21T14:43:28Z"
>   hostIP: 192.168.123.221
>   phase: Running
>   podIP: 172.24.3.89
>   podIPs:
>   - ip: 172.24.3.89
>   qosClass: BestEffort
>   startTime: "2022-04-21T14:43:25Z"
> ```
> 
> But then the net-attach-def does have that default route:
> 
> ```
> [root@openshift-jumpserver-0 ~]# oc get network-attachment-definition egress-router-cni-nad -o yaml
> apiVersion: k8s.cni.cncf.io/v1
> kind: NetworkAttachmentDefinition
> metadata:
>   annotations:
>     release.openshift.io/version: 4.9.28
>   creationTimestamp: "2022-04-21T14:43:28Z"
>   generation: 1
>   name: egress-router-cni-nad
>   namespace: sriov-testing
>   ownerReferences:
>   - apiVersion: network.operator.openshift.io/v1
>     controller: true
>     kind: EgressRouter
>     name: egress-router-redirect
>     uid: 37500600-ce3b-4190-8f21-d5e5bf65aada
>   resourceVersion: "1050659"
>   uid: 5d67e372-4117-413f-a3fa-5960ec32ac9f
> spec:
>   config: |-
>     { "cniVersion": "0.4.0", "type": "egress-router", "name": "egress-router-cni-nad", "ip": { "addresses": [ "192.168.123.180/24" ], "destinations": ["8080 TCP 192.168.123.11 80"],
>     "gateway": "192.168.123.1" }, "log_file": "/tmp/egress-router-log", "log_level": "debug" }
> ```
> 
> That's from here: https://github.com/openshift/cluster-network-operator/blob/1cb4553ed3e4d683c8bc82a0e3ad002140104bbc/bindata/egress-router/000-nad.yaml
> 
> I think the network operator is responsible for all of that setup, and so the fix would have to go there (create correct net-attach-defs or figure out something else).
> 
> I'll look further into this after the meeting.

CNO doesn't know about cluster subnet nor svc I spoke with @trozet about this and he told me similar hack was done for core-dns in ovnk

File: go-controller/pkg/ovn/pods.go [hunk 1]
```diff
@@ -275,6 +275,12 @@ func (oc *Controller) addRoutesGatewayIP(pod *kapi.Pod, podAnnotation *util.PodA
 		if isIPv6 {
 			otherDefaultRoute = otherDefaultRouteV6
 		}
+
+		// HACK Alert for egress router CNI we need to have route to cluster subnets and service CIDR to be able to connect
+		// to pods on different nodes.
+		if len(pod.Labels) != 0 && pod.Labels["app"] == "egress-router-cni" {
+			otherDefaultRoute = true
```
[the remark this answers is not in the document]
Comment (msherif1234) [reply go-controller/pkg/ovn/pods.go:284]: added the HACK marker

**Review by andreaskaris — CHANGES_REQUESTED** [review]
We're definitely looking at the wrong component here.

I think there's a bug in how network manager sets this up, and that's why we're hitting this. Possibly OVN Kube handled things differently a year ago when this was all introduced, but now, the combo of CNO and egress-router-cni does some messy, unnecessary stuff.

go-controller/vendor/github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/utils/net-attach-def.go
~~~
176 // ParseNetworkAnnotation parses actual annotation string and get NetworkSelectionElement
177 func ParseNetworkAnnotation(podNetworks, defaultNamespace string) ([]*v1.NetworkSelectionElement, error) {
178         var networks []*v1.NetworkSelectionElement
179 
180         if podNetworks == "" {
181                 return nil, fmt.Errorf("parsePodNetworkAnnotation: pod annotation not having \"network\" as key")
182         }
183 
184         if strings.IndexAny(podNetworks, "[{\"") >= 0 {
185                 if err := json.Unmarshal([]byte(podNetworks), &networks); err != nil {
186                         return nil, fmt.Errorf("parsePodNetworkAnnotation: failed to parse pod Network Attachment Selection Annotation JSON format: %v", err)
187                 }
~~~

Here: go-controller/vendor/github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1/types.go
~~~
101         // GatewayRequest contains default route IP address for the pod
102         GatewayRequest []net.IP `json:"default-route,omitempty"`
~~~

**Now, this here is super important:**
The CNI field is indeed called "gateway". 
But the multus field is called "default-route": https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/pkg/types/types.go#L102
Also check here: https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit#heading=h.hylsbqoj5fxd

The egress-router-cni however expects "gateway": https://github.com/openshift/egress-router-cni/tree/9703590c71da9bf6ba7fed7335723a5b67a9b591

However, the NetworkOperator sets this up with the wrong field, `gateway`:
~~~
[root@openshift-jumpserver-0 ~]# oc get net-attach-def egress-router-cni-nad -o yaml
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  annotations:
    release.openshift.io/version: 4.9.28
  creationTimestamp: "2022-04-21T14:43:28Z"
  generation: 1
  name: egress-router-cni-nad
  namespace: sriov-testing
  ownerReferences:
  - apiVersion: network.operator.openshift.io/v1
    controller: true
    kind: EgressRouter
    name: egress-router-redirect
    uid: 37500600-ce3b-4190-8f21-d5e5bf65aada
  resourceVersion: "1050659"
  uid: 5d67e372-4117-413f-a3fa-5960ec32ac9f
spec:
  config: |-
    { "cniVersion": "0.4.0", "type": "egress-router", "name": "egress-router-cni-nad", "ip": { "addresses": [ "192.168.123.180/24" ], "destinations": ["8080 TCP 192.168.123.11 80"],
    "gateway": "192.168.123.1" }, "log_file": "/tmp/egress-router-log", "log_level": "debug" }
~~~

If the NetworkOperator set this up correctly, then OVNKubernetes would take care of removing the wrong default route, and of adding all of the other necessary routes:
~~~
197 func (oc *Controller) addRoutesGatewayIP(pod *kapi.Pod, podAnnotation *util.PodAnnotation, nodeSubnets []*net.IPNet) error {
198         // if there are other network attachments for the pod, then check if those network-attachment's
199         // annotation has default-route key. If present, then we need to skip adding default route for
200         // OVN interface
201         networks, err := util.GetK8sPodAllNetworks(pod)
202         if err != nil {
203                 return fmt.Errorf("error while getting network attachment definition for [%s/%s]: %v",
204                         pod.Namespace, pod.Name, err)
205         }       
206         otherDefaultRouteV4 := false
207         otherDefaultRouteV6 := false
208         for _, network := range networks {
209                 for _, gatewayRequest := range network.GatewayRequest {
210                         if utilnet.IsIPv6(gatewayRequest) {
211                                 otherDefaultRouteV6 = true
212                         } else {
213                                 otherDefaultRouteV4 = true
214                         }
215                 }
 NORMAL  ᚠ egress-firewall-acl-3  go-controller/pkg/ovn/pods.go  
~~~
... meaning in turn that the proposed hack here is not needed, because it basically does the exact same thing.

The bug in the CNO is here:
https://github.com/openshift/cluster-network-operator/blob/1cb4553ed3e4d683c8bc82a0e3ad002140104bbc/bindata/egress-router/000-nad.yaml#L22
~~~
spec:
  config: '{
    "cniVersion": "0.4.0",
    "type": "egress-router",
    "name": "egress-router-cni-nad",
    "ip": {
      "addresses": [
        "{{.Addresses}}"
        ],
      "destinations": {{.AllowedDestinations}},
      {{ $fallbackip := .FallbackIP}} {{ if ne $fallbackip "" }}
        "fallbackIP": "{{$fallbackip}}",
      {{ end }}
      "gateway": "{{.Gateway}}"                # <----------- this must be default-route
        },
      "log_file": "/tmp/egress-router-log",
      "log_level": "debug"
    }'
~~~

And because of that tiny bug / typo, we then pull a whole lot of weird tricks out of our hat to make this work:
https://github.com/openshift/egress-router-cni/blob/9703590c71da9bf6ba7fed7335723a5b67a9b591/README.md#routing

I **suppose** that multus just passes through all of the parameters that it gets to the CNI plugin, and the CNI parameters IIRC can be any random elements (aside from a bunch of standardized elements).
I don't understand multus and the plugin chaining that it does :-/ When we have a net-attach-def and ovn-kubernetes, which plugin comes first? OVN-K, and then the plugin from
the net-attach-def? From what I'm seeing, it's OVN-K first, then egress-router-cni after that.

We parse the gateway here in egress-router-cni pkg/macvlan/macvlan.go, and that RouteAdd and RouteDelete part is obviously not necessary because OVN-Kubernetes should already take care of it if we used `default-route` inside the net-attach-def:
~~~
341 func macvlanCmdAdd(args *skel.CmdArgs) error {
342         n, err := loadNetConf(&types.ClusterConf{}, args.StdinData)
343         logging.Debugf("Called CNI ADD")
344         if err != nil {
345                 return err
346         }
347         logging.Debugf("Gateway: %s", n.IP.Gateway)
348         logging.Debugf("IP Source Addresses: %s", n.IP.Addresses)
349         logging.Debugf("IP Destinations: %v", n.IP.Destinations)
350 
351         netns, err := ns.GetNS(args.Netns)
352         if err != nil {
353                 return fmt.Errorf("failed to open netns %q: %v", netns, err)
354         }
355         defer netns.Close()
356 
357         macvlanInterface, err := createMacvlan(n, args.IfName, netns)
358         if err != nil {
359                 return err
360         }
(...)
447                 // Delete default route
448                 var routes []netlink.Route
449                 if isIPv6 {
450                         routes, _ = netlink.RouteList(existingLink, netlink.FAMILY_V6)
451                 } else {
452                         routes, _ = netlink.RouteList(existingLink, netlink.FAMILY_V4)
453                 }
454 
455                 for _, r := range routes {
456                         if r.Dst == nil {
457                                 if err := netlink.RouteDel(&r); err != nil {
458                                         logging.Errorf("failed to delete existing default route : %v", err)
459                                         return fmt.Errorf("failed to delete existing default route : %v", err)
460                                 }
461                                 logging.Debugf("deleted default route %v", r)
462                         }
463                 }
464 
465                 // Create new default route
466                 newDefaultRoute := netlink.Route{
467                         LinkIndex: macvlanLink.Attrs().Index,
468                         Dst:       nil,
469                         Gw:        gw,
470                 }
471 
472                 if err := netlink.RouteAdd(&newDefaultRoute); err != nil {
473                         // Check if we already have route installed
474                         if !os.IsExist(err) {
475                                 logging.Errorf("failed to add new default route, gw %v : %v", gw, err)
476                                 return fmt.Errorf("failed to add new default route, gw %v : %v", gw, err)
477                         }
478                         logging.Debugf("Use existing route with gateway %v", gw)
479                 } else {
480                         logging.Debugf("Added new default route with gateway %v", gw)
481                 }

(...)
 NORMAL  ᚠ master  pkg/macvlan/macvlan.go  
~~~

So, the fix here is 2fold.

a) Fix the annotation that ClusterNetworkManager creates.
b) Fix the code in egress-router-cni to only do the necessary work (as in, if the def route is where it should be, do not delete it and keep the other routes).
c) Possibly also update egress-router-cni to use `default-route` instead of `gateway` **if** that's compatible with OpenShiftSDN.
d) Check for any conflicts with OpenShiftSDN - does OpenShift SDN use `default-route` or `gateway` or neither? Possibly duplicate annotations to use `default-route` and `gateway` if we get into conflicts otherwise with OpenShift-SDN or align 

Perhaps we can solve it all with an "ugly hack" and just add both  `default-route` and `gateway` in the CNO and see what that gives???

**Comment by msherif1234** [comment]
> We're definitely looking at the wrong component here.
> 
> I think there's a bug in how network manager sets this up, and that's why we're hitting this. Possibly OVN Kube handled things differently a year ago when this was all introduced, but now, the combo of CNO and egress-router-cni does some messy, unnecessary stuff.
> 
> go-controller/vendor/github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/utils/net-attach-def.go
> 
> ```
> 176 // ParseNetworkAnnotation parses actual annotation string and get NetworkSelectionElement
> 177 func ParseNetworkAnnotation(podNetworks, defaultNamespace string) ([]*v1.NetworkSelectionElement, error) {
> 178         var networks []*v1.NetworkSelectionElement
> 179 
> 180         if podNetworks == "" {
> 181                 return nil, fmt.Errorf("parsePodNetworkAnnotation: pod annotation not having \"network\" as key")
> 182         }
> 183 
> 184         if strings.IndexAny(podNetworks, "[{\"") >= 0 {
> 185                 if err := json.Unmarshal([]byte(podNetworks), &networks); err != nil {
> 186                         return nil, fmt.Errorf("parsePodNetworkAnnotation: failed to parse pod Network Attachment Selection Annotation JSON format: %v", err)
> 187                 }
> ```
> 
> Here: go-controller/vendor/github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1/types.go
> 
> ```
> 101         // GatewayRequest contains default route IP address for the pod
> 102         GatewayRequest []net.IP `json:"default-route,omitempty"`
> ```
> 
> **Now, this here is super important:** The CNI field is indeed called "gateway". But the multus field is called "default-route": https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/pkg/types/types.go#L102 Also check here: https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit#heading=h.hylsbqoj5fxd
> 
> The egress-router-cni however expects "gateway": https://github.com/openshift/egress-router-cni/tree/9703590c71da9bf6ba7fed7335723a5b67a9b591
> 
> However, the NetworkOperator sets this up with the wrong field, `gateway`:
> 
> ```
> [root@openshift-jumpserver-0 ~]# oc get net-attach-def egress-router-cni-nad -o yaml
> apiVersion: k8s.cni.cncf.io/v1
> kind: NetworkAttachmentDefinition
> metadata:
>   annotations:
>     release.openshift.io/version: 4.9.28
>   creationTimestamp: "2022-04-21T14:43:28Z"
>   generation: 1
>   name: egress-router-cni-nad
>   namespace: sriov-testing
>   ownerReferences:
>   - apiVersion: network.operator.openshift.io/v1
>     controller: true
>     kind: EgressRouter
>     name: egress-router-redirect
>     uid: 37500600-ce3b-4190-8f21-d5e5bf65aada
>   resourceVersion: "1050659"
>   uid: 5d67e372-4117-413f-a3fa-5960ec32ac9f
> spec:
>   config: |-
>     { "cniVersion": "0.4.0", "type": "egress-router", "name": "egress-router-cni-nad", "ip": { "addresses": [ "192.168.123.180/24" ], "destinations": ["8080 TCP 192.168.123.11 80"],
>     "gateway": "192.168.123.1" }, "log_file": "/tmp/egress-router-log", "log_level": "debug" }
> ```
> 
> If the NetworkOperator set this up correctly, then OVNKubernetes would take care of removing the wrong default route, and of adding all of the other necessary routes:
> 
> ```
> 197 func (oc *Controller) addRoutesGatewayIP(pod *kapi.Pod, podAnnotation *util.PodAnnotation, nodeSubnets []*net.IPNet) error {
> 198         // if there are other network attachments for the pod, then check if those network-attachment's
> 199         // annotation has default-route key. If present, then we need to skip adding default route for
> 200         // OVN interface
> 201         networks, err := util.GetK8sPodAllNetworks(pod)
> 202         if err != nil {
> 203                 return fmt.Errorf("error while getting network attachment definition for [%s/%s]: %v",
> 204                         pod.Namespace, pod.Name, err)
> 205         }       
> 206         otherDefaultRouteV4 := false
> 207         otherDefaultRouteV6 := false
> 208         for _, network := range networks {
> 209                 for _, gatewayRequest := range network.GatewayRequest {
> 210                         if utilnet.IsIPv6(gatewayRequest) {
> 211                                 otherDefaultRouteV6 = true
> 212                         } else {
> 213                                 otherDefaultRouteV4 = true
> 214                         }
> 215                 }
>  NORMAL  ᚠ egress-firewall-acl-3  go-controller/pkg/ovn/pods.go  
> ```
> 
> ... meaning in turn that the proposed hack here is not needed, because it basically does the exact same thing.
> 
> The bug in the CNO is here: https://github.com/openshift/cluster-network-operator/blob/1cb4553ed3e4d683c8bc82a0e3ad002140104bbc/bindata/egress-router/000-nad.yaml#L22
> 
> ```
> spec:
>   config: '{
>     "cniVersion": "0.4.0",
>     "type": "egress-router",
>     "name": "egress-router-cni-nad",
>     "ip": {
>       "addresses": [
>         "{{.Addresses}}"
>         ],
>       "destinations": {{.AllowedDestinations}},
>       {{ $fallbackip := .FallbackIP}} {{ if ne $fallbackip "" }}
>         "fallbackIP": "{{$fallbackip}}",
>       {{ end }}
>       "gateway": "{{.Gateway}}"                # <----------- this must be default-route
>         },
>       "log_file": "/tmp/egress-router-log",
>       "log_level": "debug"
>     }'
> ```
> 
> And because of that tiny bug / typo, we then pull a whole lot of weird tricks out of our hat to make this work: https://github.com/openshift/egress-router-cni/blob/9703590c71da9bf6ba7fed7335723a5b67a9b591/README.md#routing
> 
> I **suppose** that multus just passes through all of the parameters that it gets to the CNI plugin, and the CNI parameters IIRC can be any random elements (aside from a bunch of standardized elements). I don't understand multus and the plugin chaining that it does :-/ When we have a net-attach-def and ovn-kubernetes, which plugin comes first? OVN-K, and then the plugin from the net-attach-def? From what I'm seeing, it's OVN-K first, then egress-router-cni after that.
> 
> We parse the gateway here in egress-router-cni pkg/macvlan/macvlan.go, and that RouteAdd and RouteDelete part is obviously not necessary because OVN-Kubernetes should already take care of it if we used `default-route` inside the net-attach-def:
> 
> ```
> 341 func macvlanCmdAdd(args *skel.CmdArgs) error {
> 342         n, err := loadNetConf(&types.ClusterConf{}, args.StdinData)
> 343         logging.Debugf("Called CNI ADD")
> 344         if err != nil {
> 345                 return err
> 346         }
> 347         logging.Debugf("Gateway: %s", n.IP.Gateway)
> 348         logging.Debugf("IP Source Addresses: %s", n.IP.Addresses)
> 349         logging.Debugf("IP Destinations: %v", n.IP.Destinations)
> 350 
> 351         netns, err := ns.GetNS(args.Netns)
> 352         if err != nil {
> 353                 return fmt.Errorf("failed to open netns %q: %v", netns, err)
> 354         }
> 355         defer netns.Close()
> 356 
> 357         macvlanInterface, err := createMacvlan(n, args.IfName, netns)
> 358         if err != nil {
> 359                 return err
> 360         }
> (...)
> 447                 // Delete default route
> 448                 var routes []netlink.Route
> 449                 if isIPv6 {
> 450                         routes, _ = netlink.RouteList(existingLink, netlink.FAMILY_V6)
> 451                 } else {
> 452                         routes, _ = netlink.RouteList(existingLink, netlink.FAMILY_V4)
> 453                 }
> 454 
> 455                 for _, r := range routes {
> 456                         if r.Dst == nil {
> 457                                 if err := netlink.RouteDel(&r); err != nil {
> 458                                         logging.Errorf("failed to delete existing default route : %v", err)
> 459                                         return fmt.Errorf("failed to delete existing default route : %v", err)
> 460                                 }
> 461                                 logging.Debugf("deleted default route %v", r)
> 462                         }
> 463                 }
> 464 
> 465                 // Create new default route
> 466                 newDefaultRoute := netlink.Route{
> 467                         LinkIndex: macvlanLink.Attrs().Index,
> 468                         Dst:       nil,
> 469                         Gw:        gw,
> 470                 }
> 471 
> 472                 if err := netlink.RouteAdd(&newDefaultRoute); err != nil {
> 473                         // Check if we already have route installed
> 474                         if !os.IsExist(err) {
> 475                                 logging.Errorf("failed to add new default route, gw %v : %v", gw, err)
> 476                                 return fmt.Errorf("failed to add new default route, gw %v : %v", gw, err)
> 477                         }
> 478                         logging.Debugf("Use existing route with gateway %v", gw)
> 479                 } else {
> 480                         logging.Debugf("Added new default route with gateway %v", gw)
> 481                 }
> 
> (...)
>  NORMAL  ᚠ master  pkg/macvlan/macvlan.go  
> ```
> 
> So, the fix here is 2fold.
> 
> a) Fix the annotation that ClusterNetworkManager creates. b) Fix the code in egress-router-cni to only do the necessary work (as in, if the def route is where it should be, do not delete it and keep the other routes). c) Possibly also update egress-router-cni to use `default-route` instead of `gateway` **if** that's compatible with OpenShiftSDN. d) Check for any conflicts with OpenShiftSDN - does OpenShift SDN use `default-route` or `gateway` or neither? Possibly duplicate annotations to use `default-route` and `gateway` if we get into conflicts otherwise with OpenShift-SDN or align
> 
> Perhaps we can solve it all with an "ugly hack" and just add both `default-route` and `gateway` in the CNO and see what that gives???

1st its extremenlly hard to follow those large comments and comment on them
- so I understood now that Multus could/may off load some of the code in the egress cni where it populate the default route 

however that has nothing to do with the issue we have in hand here we still need to populate pods subnet route to access pods on different node, cni doesn't have anything with annotation 
gateway is what we have in the defined apis so there no wrong field here 
probably when use Multus NAD we probably don't need to render anything in gateway field

**Comment by andreaskaris** [comment]
The correct place to fix this would be: https://github.com/openshift/cluster-network-operator/pull/1390
